builtin/credential/aws-iam/path_login.go

```diff
@@ -0,0 +1,376 @@
+	if err != nil {
+		return logical.ErrorResponse("Error getting configuration"), nil
+	}
+
+	if config.HeaderValue != "" {
+		ok, msg := ensureVaultHeaderValue(headers, parsedUrl, config.HeaderValue)
+		if !ok {
+			return logical.ErrorResponse(fmt.Sprintf("Error validating %s header: %s", magicVaultHeader, msg)), nil
+		}
+	}
+
+	endpoint := config.Endpoint
+	if endpoint == "" {
+		endpoint = "https://sts.amazonaws.com"
+	}
+
+	clientArn, err := submitCallerIdentityRequest(method, endpoint, parsedUrl, body, headers)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("Error making upstream request: %s", err)), nil
+	}
+	canonicalArn, principalName, err := parseIamArn(clientArn)
+	if err != nil {
+		return logical.ErrorResponse("Unrecognized IAM principal type"), nil
+	}
+
+	roleName := data.Get("role").(string)
+	if roleName == "" {
+		roleName = principalName
+	}
+
+	roleEntry, err := b.lockedAWSRole(req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return logical.ErrorResponse(fmt.Sprintf("entry for role %s not found", roleName)), nil
+	}
+
+	if roleEntry.BoundIamPrincipal != canonicalArn {
+		return logical.ErrorResponse(fmt.Sprintf("IAM Principal '%s' does not belong to the role '%s'", clientArn, roleName)), nil
+	}
+
+	resp := &logical.Response{
+		Auth: &logical.Auth{
+			Policies: roleEntry.Policies,
+			Metadata: map[string]string{
+				"client_arn":    clientArn,
+				"canonical_arn": canonicalArn,
+			},
+			InternalData: map[string]interface{}{
+				"role_name": roleName,
+			},
+			DisplayName: principalName,
+			LeaseOptions: logical.LeaseOptions{
+				Renewable: true,
+				TTL:       roleEntry.TTL,
+			},
+		},
+	}
+
+	shortestTTL := b.System().DefaultLeaseTTL()
+	if roleEntry.TTL > time.Duration(0) && roleEntry.TTL < shortestTTL {
+		shortestTTL = roleEntry.TTL
+	}
+
+	maxTTL := b.System().MaxLeaseTTL()
+	if roleEntry.MaxTTL > time.Duration(0) && roleEntry.MaxTTL < maxTTL {
+		maxTTL = roleEntry.MaxTTL
+	}
+
+	if shortestTTL > maxTTL {
+		resp.AddWarning(fmt.Sprintf("Effective TTL of %q exceeded the effective max_ttl of %q; TTL value is capped accordingly", (shortestTTL / time.Second).String(), (maxTTL / time.Second).String()))
+		shortestTTL = maxTTL
+	}
+
+	resp.Auth.TTL = shortestTTL
+
+	return resp, nil
+}
+
+func (b *backend) pathLoginRenew(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	canonicalArn := req.Auth.Metadata["canonical_arn"]
+	if canonicalArn == "" {
+		return nil, fmt.Errorf("unable to retrieve canonical ARN from metadata during renewal")
+	}
+
+	roleName := req.Auth.InternalData["role_name"].(string)
+	if roleName == "" {
+		return nil, fmt.Errorf("Error retrieving role_name during renewal")
+	}
+	roleEntry, err := b.lockedAWSRole(req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return nil, fmt.Errorf("role entry not found")
+	}
+
+	if roleEntry.BoundIamPrincipal != canonicalArn {
```

Comment:
That's a good question. The only thing Vault sees, by design, is the access key and the session token (if it's an STS token). There's no good way, in AWS, to determine which IAM principal an access key belongs to. The only real way of doing that is by calling the `GetCallerIdentity` method, which requires the secret key in order to sign the request, and by design, Vault never sees the secret key. Further, as @copumpkin states, the request itself will expire within 15 minutes, so it's not feasible to really store it.

It is possible, as @copumpkin states, to add a date constraint to an IAM role policy to cut off access for all session tokens that were issued before a certain date. However, _as best as I can tell_, `GetCallerIdentity` doesn't actually do any authorization checks, and so this trick won't work for denying future `GetCallerIdentity` requests.

If a Vault admin is concerned by this, then I think the right solution is some combination of low token TTL (so forcing a periodic renewal) and/or low max token TTL or just disabling renewal altogether (so forcing a revalidation that the client has valid IAM creds).

@jefferai -- in your most recent comment, you're quoting a section about `GetSessionToken` and not `GetCallerIdentity`. And, on your confusion, in my experience, I think that's inconsistent documentation standards. In this context, I think that the "account credentials" is credentials for the root user (which should almost never be used), IAM users are long-lived, static creds that we're desperately trying to avoid building a dependency on, and then you have short-lived creds that we like because they expire automatically. So, yes, the user could submit a new signed request with a renewal, but I think that breaks the paradigm of a "renewal" (which is just the client presenting the original token, without additional authenticating information, and asking for the lifetime to be extended, right?). I think that, if you want the user to have to present a new signed request, the easiest and best way to do that would be to just set a low TTL and max TTL (e.g., disallow renewals) and force the client to get a new authentication token.

This would probably be good to capture in the docs. @jefferai -- let me know if this makes sense to you and I'll add it.